### PR TITLE
Relax GDScript name lookup for tool scripts

### PR DESCRIPTION
**Pull request by ghost**
In scripts that are `tool` as well as runtime scripts, accessing singletons is broken, i.e.:

```
tool

func my_func():
	return my_singleton
```

will fail to even compile in the editor, as `my_singleton` is only defined at runtime (thus there's no way to fix this in the script).

The alternative, to use `get_node("/root/my_singleton")` doesn't work either if your code needs to run outside of the main scene tree (`get_node` will fail with "Attempt to get SceneTree while node is not in the active tree.", while the global `my_singleton` would be accessible there).

~~Allowing singletons in the editor would be one fix, but that is very complex. Extending `get_node` for a special case is complex too. I suggest we simply relax name lookups in `tool` scripts (only there): if a symbol is not defined there, it just evaluates to `null` (instead of a compile error). This allows users to add code to properly handle the editor case (e.g. `if my_singleton:` - note that `if Engine.is_editor_hint()` won't work due to the compile error) while keeping the non-`get_node` code for the runtime case that works everywhere.~~ 

See https://github.com/godotengine/godot/pull/17511

### CODE REVIEW

**Comment by Zylann** [comment]
Would it be possible for singletons to work in tool mode though? I've been needing this recently but had to stuff everything in my custom node instead (which then will allocate copies of that for every instance)

**Comment by Zylann** [comment]
Supporting singletons in editor would be really nice... I've been needing this recently but had to stuff everything in my custom node instead (which then will allocate suboptimal copies of that for every instance)

File: modules/gdscript/gdscript_compiler.cpp [hunk 1]
```diff
@@ -278,6 +278,11 @@ int GDScriptCompiler::_parse_expression(CodeGen &codegen, const GDScriptParser::
 				return idx | (GDScriptFunction::ADDR_TYPE_GLOBAL << GDScriptFunction::ADDR_BITS); //argument (stack root)
 			}
 
+			if (codegen.script->tool) {
```
Comment (bojidar-bg) [inline modules/gdscript/gdscript_compiler.cpp]: Can you make this check stricter by checking if the singleton is defined in ProjectSettings?

**Comment by vnen** [comment]
I created a PR to support autoloads in-editor (#18545) which would solve this use case in a better way, since you could actually call functions from singletons in the editor (if such singleton is also a tool script).

**Comment by akien-mga** [comment]
Superseded by #18545.

**Pull request closed without merging**